### PR TITLE
Feat/versioning

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
Automate the versioning scheme using `bumpversion`

## Current workflow
The current versioning scheme relies on multiple sources of "truth" in that there are several version tags located throughout the codebase:
1. [`__init__.py`](https://github.com/chemprop/chemprop/tree/master/chemprop/__init__.py)
2. [**`_version.py`**](https://github.com/chemprop/chemprop/tree/master/chemprop/_version.py) <-- intendened **single** source of truth
3. [`setup.py`](https://github.com/chemprop/chemprop/blob/master/setup.py)
4. [`conf.py`](https://github.com/chemprop/chemprop/blob/master/docs/source/conf.py)

(1) and (2) are supposedly linked at runtime, (2) and (3) are linked at build time, but (4) is completely independent. This is undesired for a few reasons:
- there are two sources of runtime version truth that are only linked upon initial import: `chemprop.__version__` and `chemprop._version.__version__`. These can possibly become unlinked during runtime due to python's mutable nature, leading a client to doubt which one to trust. I can't imagine this is actually a big issue (because why would client code depend heavily on version throughout program runtime?) but it is an _undesirable_ trait nonetheless.
- the middle two are linked at build time and package distribution with the following snippet:
```python
# Load version number
__version__ = None

src_dir = os.path.abspath(os.path.dirname(__file__))
version_file = os.path.join(src_dir, 'chemprop', '_version.py')

with open(version_file, encoding='utf-8') as fd:
    exec(fd.read())
```
This snippet implicitly links the version to the current source distribution. Typically, this would induce a build-time dependency on the package itself, but the snippet gets around that by `exec()`ing the raw source file to obtain the current version from the intended single source of truth. However, using `exec` in code is widely regarded as bad practice and introduces potential security vulnerabilities, so it should best be avoided. A simple workaround to this would be to parse the source file for the version information and construct a version string from there, but ideally, the version should be package metadata that has its SSOT **external** to the package itself
- the docs version number must be manually changed every time the version is changed. This is mostly an inconvenience, but it's one that can be avoided.
- why do we need a special `_version` module anyway? What does a whole module do that can't be accomplished with a simple `__version__` attribute of the main module?

## Changes / New workflow
This PR introduces a `bumpversion` dev dependency to automate all of these versioning pains. Now, you can simply run `bumpversion major|minor|patch` and the version strings in all of the above files will be changed. This PR also removes the `_version` module entirely. We can add it back at a later date if we need its utility, and that will be a simple change (moving `__version__` from `__init__.py` to `_version.py` and importing the version tag from there- like how it's done now.)

## Questions
This PR is intended to be the first a series of build and development changes to chemprop that both ease and automate the entire process. There are a few points to consider with this PR:
- This PR has removed version suffixes (e.g., `a1`, `b1`, etc.) We can still have these using `bumpversion` with some amount of hacking, but it's not as straightforward as it previously was. I can't imagine this is a real issue, and if it is needed, `bumpversion` does support this via its configuration options
- the default config is to tag a version bump. This is disabled via the `--no-tag` flag when version bumping. It could be tedious to do this all the time, but I would imagine most times that we version bump we will want a tag (the exception being patches.) This can be changed if we think it's better to default to no tagging.

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by kevingreenman** [comment]
I'm not as familiar with the versioning process/requirements (especially as they relate to our commitments to MLPDS), so I defer to @cjmcgill, but this seems like a logical change to me.

**Comment by mliu49** [comment]
I think the Python packaging guide has a good overview of various options: https://packaging.python.org/en/latest/guides/single-sourcing-package-version/

I'm not sure how necessary it is to introduce a dependency for versioning. It seems to me that the only non-automatic version number right now is in conf.py, but that could be easily changed to import `__version__` instead (since sphinx imports everything anyway, and [sets up sys.path](https://github.com/chemprop/chemprop/blob/master/docs/source/conf.py#L15) itself).

I also don't quite understand why the SSOT needs to be outside of the package. However, an option that fits that and eliminates eval would be a basic text file containing the version number, which could potentially be read independently by `__init__.py` and `setup.py`.

It looks like bumpversion also automates creating a git commit and tag, but we might not be able to take advantage of that if it doesn't fit the current workflow, e.g. if version number bump happens in a feature branch and then the tag goes on the master branch.

(Also, less important, but bumpversion is unmaintained, so we would probably want to use bump2version.)

**Comment by davidegraff** [comment]
`bumpversion` on PyPI has been rerouted to the maintained `bump2version`, so we don't have to worry about using unmaintained software.

Package version is one of those weird things that it seems to be just standard practice where the information is stored *inside* the package itself, but it is firmly *metadata*. Ideally, we wouldn't be mixing these responsibilities. After all, we don't store package descriptions, authors, homepages, etc. inside the package. While we *can* version our package using information contained inside the package source, my understanding is that this is a practice that's fallen out of favor and is no longer used in a number of major packages (numpy, scipy, pytorch, tensorflow.) Rather, their package version info is _generated_ in their source tree at build-time rather than _determined_ from their source tree at build time. We could implement something like with great effort, or just use `bumpversion` to automatically handle the updating of all sources of `__version__` inside our codebase whenever it's used.

As far as workflow goes, I don't know what the version bumping scheme currently we currently use is. With `bumpversion` we can create an action that runs on `master` when we signal it to and bumps the appropriate version along with tagging it. Alternative, and less complex, is to just offload the responsibility to one person to run `bumpversion` on `master`.

**Comment by cjmcgill** [comment]
I think that if you add the github action to upload to pypi as part of this pr or a separate closely following pr, then I'd be on board with this as a general streamlining of the version and release process.

File: chemprop/__init__.py [hunk 1]
```diff
@@ -16,4 +16,6 @@
 import chemprop.spectra_utils
 import chemprop.hyperopt_utils
 
-from chemprop._version import __version__
+# from chemprop._version import __version__
+
+__version__ = "1.4.1"
```
Comment (cjmcgill) [inline chemprop/__init__.py:21]: We should have a newline at the end right?

**Review by cjmcgill — APPROVED** [review]
Looks good

File: chemprop/__init__.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/__init__.py:21]: i guess that would be proper

**Pull request merged**